capstone/src/main/java/com/google/sps/Objects/QuadTree.java

```diff
@@ -0,0 +1,192 @@
+package com.google.sps.Objects;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.PriorityQueue;
+import java.util.Comparator;
+
+import com.google.sps.Objects.Location;
+import com.google.sps.Objects.Coordinate;
+import com.google.sps.Objects.BoundingBox;
+
+public class QuadTree {
+  private final int NODE_CAPACITY;
+  private List<QuadTree> children;
+  private List<Location> locations;
+  private QuadTree topLeftTree, topRightTree, botLeftTree, botRightTree; 
+  private BoundingBox bounds;
+
+  // For keeping track of splits and number of points within each quadtree
+  public int level;
+  public int numPoints;
+
+  public QuadTree(int NODE_CAPACITY, int level, BoundingBox bounds) {
+    this.NODE_CAPACITY = NODE_CAPACITY;
+    this.bounds = bounds;
+    this.children = new ArrayList<QuadTree>();
+    this.locations = new ArrayList<Location>();
+    this.level = level;
+    this.numPoints = 0;
+  
+    this.topLeftTree  = null; 
+    this.topRightTree = null; 
+    this.botLeftTree  = null; 
+    this.botRightTree = null; 
+  }
+
+  public List<QuadTree> getChildren() {
+    return this.children;
+  }
+
+  public List<Location> getLocations() {
+    return this.locations;
+  }
+
+  public QuadTree getTopLeftTree() {
+    return topLeftTree;
+  }
+
+  public QuadTree getTopRightTree() {
+    return topRightTree;
+  }
+
+  public QuadTree getBottomLeftTree() {
+    return botLeftTree;
+  }
+
+  public QuadTree getBottomRightTree() {
+    return botRightTree;
+  }
+
+  public double euclidianDistance(Location loc1, Location loc2) {
+    return Math.sqrt(Math.pow(loc1.getLatitude() - loc2.getLatitude(), 2) + Math.pow(loc1.getLongitude() - loc2.getLongitude(), 2));
+  }
+
+  public boolean insert(Location node) { 
+    if (node == null || !bounds.containsPoint(node)) return false; 
+
+    if (locations.size() < NODE_CAPACITY) {
+      locations.add(node);
+      numPoints++;
+      return true;
+    }
+
+    // Exceeded the capacity so split into four quadrants
+    if (topLeftTree == null) {
+      split();
+    }
+
+    // Add point to correct quadrant 
+    for (QuadTree qt: children) {
+      if (qt.insert(node)) {
+        numPoints ++;
+        return true;
+      }
+    }
+
+    // Node is not in this bounding box
+    return false;
+  }
+
+  void split() {
```

Comment:
Can you add comments to these methods to describe what they are all doing? There is a lot of math happening so it would also be good to make sure it's documented well in your design doc

Reply:
Add some comments, let me know if I can make anything more clear!